modules/ruxhal/src/platform/aarch64_common/ivc.rs

```diff
@@ -0,0 +1,218 @@
+         info!("Error sending message: {}", e);
+     }
+ 
+     // Close the connection
+     if let Err(e) = conn.close() {
+         info!("Error closing connection: {}", e);
+     }
+ }
+ 
+ pub struct Connection<'a> {
+     // Reference to the allocator
+     allocator: &'a GlobalAllocator, 
+     ivc_info_ptr: *mut IvCInfo,
+     control_table_ptr: *mut ControlTable,
+ }
+ 
+ impl<'a> Connection<'a> {
+     
+     pub fn new() -> Self {
+         // Get the reference to the global allocator
+         let allocator = global_allocator(); 
+         debug!("Connection created.");
+         Connection {
+             allocator,
+             ivc_info_ptr: ptr::null_mut(),
+             control_table_ptr: ptr::null_mut(),
+         }
+     }
+ 
+     pub fn connect(&mut self, communication_zone: usize) -> Result<(), &'static str> {
+         let alloc_size = HVISOR_HC_IVC_INFO_SIZE;
+         let align = HVISOR_HC_IVC_INFO_ALIGN;
+         let layout = Layout::from_size_align(alloc_size, align).unwrap();
+         
+         let ptr = self.allocator.alloc(layout).expect("Memory allocation failed!");
+         self.ivc_info_ptr = ptr.as_ptr() as *mut IvCInfo;
+ 
+         let vpa_ivcinfo = self.ivc_info_ptr as usize;
+         let pa_ivcinfo: usize = vpa_ivcinfo - __PA;
+ 
+         debug!("The memory address of the IVC Info: VA: 0x{:x}, IPA: 0x{:x}", vpa_ivcinfo, pa_ivcinfo);
+ 
+         ivc_hvc_call(HVISOR_HC_IVC_INFO as u32, pa_ivcinfo, HVISOR_HC_IVC_INFO_SIZE);
+         debug!("ivc_hvc_call finished.");
+ 
+         // Safety: At this point we know ivc_info_ptr is valid and allocated
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+ 
+         let control_table_ptr = (ivc_info.ivc_ct_ipas[communication_zone] + __PA as u64) as *mut ControlTable;
+         self.control_table_ptr = control_table_ptr;
+ 
+         info!("IVC Connection established.");
+         Ok(())
+     }
+ 
+     pub fn send_message(&self, message: &str) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+     
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+     
+         let control_table: &ControlTable = unsafe { &*self.control_table_ptr };
+     
+         // Suppose we are zone1, we are to send message to zone0.
+         // Therefore use the out_sec_size field of the ControlTable struct (typically 0x1000).
+         let offset = control_table.out_sec_size as u64;
+         let address: u64 = ivc_info.ivc_shmem_ipas[0] + offset + __PA as u64;
+     
+         write_to_address(address, message)?;
+         info!("Message written to shared memory: {}", message);
+     
+         // Modify the control table to inform Zone0 Linux
+         let control_table: &mut ControlTable = unsafe { &mut *self.control_table_ptr };
+         debug!("Ipi_invoke reset to inform Zone0 linux.");
+         control_table.ipi_invoke = 0x0;
```

Comment:
what is the meaning of `0x0`?

Reply:
Here are the communication requirements for hvisor.  Writing to this field triggers an interrupt notification. You can refer to the highlighted sections in [this document](https://hzl24dbxn8.feishu.cn/docx/VlEmdvM1woDxppxRwBpc7ldlnrb) for further details. Is there any better approach to write here?